esphome/components/remote_receiver/remote_receiver_esp8266.cpp

```diff
@@ -12,20 +12,61 @@ static const char *const TAG = "remote_receiver.esp8266";
 
 void IRAM_ATTR HOT RemoteReceiverComponentStore::gpio_intr(RemoteReceiverComponentStore *arg) {
   const uint32_t now = micros();
-  // If the lhs is 1 (rising edge) we should write to an uneven index and vice versa
   const uint32_t next = (arg->buffer_write_at + 1) % arg->buffer_size;
+
+  // If the lhs is 1 (rising edge) we should write to an uneven index and vice versa
   const bool level = arg->pin.digital_read();
   if (level != next % 2)
     return;
 
-  // If next is buffer_read, we have hit an overflow
-  if (next == arg->buffer_read_at)
+  // If next is buffer_read_at, we have hit an overflow
+  if (next == arg->buffer_read_at) {
+    arg->overflow = true;
+    // Reset the write_at to the last good idle position corresponding to signal level,
+    // droping some data since it's is already corrupted
+    if (arg->buffer_write_at != arg->buffer_idle_at) {
+      if (level == arg->buffer_idle_at % 2) {
+        arg->buffer_write_at = arg->buffer_idle_at;
+      } else {
+        arg->buffer_write_at = (arg->buffer_idle_at + 1) % arg->buffer_size;
+      }
+    }
+    arg->buffer[arg->buffer_write_at] = now;
     return;
+  }
 
   const uint32_t last_change = arg->buffer[arg->buffer_write_at];
   const uint32_t time_since_change = now - last_change;
-  if (time_since_change <= arg->filter_us)
+  const uint32_t prev = (arg->buffer_size + arg->buffer_write_at - 1) % arg->buffer_size;
+
+  if (time_since_change >= arg->idle_us) {
```

Comment:
Can't just do something like this?

```suggestion
    // Handle idle state
    arg->overflow = false;

    if (prev == arg->buffer_idle_at) {
      // Detected duplicated ide, drop it
      return;
    } 
    
    const uint32_t prev_delta = last_change - arg->buffer[prev];
    if (prev_delta <= arg->filter_us && prev_delta != 0 && arg->buffer[prev] != 0) {
      // If delta of the previous change is less than filter_us, we can just update the previous value
      arg->buffer_write_at = prev;
      arg->buffer[arg->buffer_write_at] = now;
      arg->buffer_idle_at = prev;
      return;
    }

    arg->buffer_idle_at = next;
```